src/functions.php

```diff
@@ -186,6 +186,17 @@ function get_encrypted_fields_from_server(string $databaseName, string $collecti
     return null;
 }
 
+/**
+ * Returns whether a given document is a valid document.
+ *
+ * This method accepts any array and object, but specifically excludes BSON
+ * PackedArray instances
+ */
+function is_document($document): bool
+{
+    return is_array($document) || (is_object($document) && ! $document instanceof PackedArray);
```

Comment:
Noting that we don't bother checking if array is a list, since APIs that take arrays as documents generally use an object cast to ensure proper BSON encoding as a document type.

Reply:
Ok, but can a list be a valid value for the commands? Can `0` be used as a field name?

If not, it may help developers if they're notified as soon as possible that the value they've passed is a list of documents, when a single document is expected.

Reply:
> Ok, but can a list be a valid value for the commands?

Strictly speaking, a list is not a valid document, even though it has been treated as such in PHPLIB/PHPC.

> Can `0` be used as a field name?

Yes. Arrays are serialised like objects in BSON, but denoted through a special type. Likewise, inserting a structure like `[0 => 'foo', 2 => 'bar']` would always yield an object with `"0"` and `"2"` as keys (since the value isn't a list).

> If not, it may help developers if they're notified as soon as possible that the value they've passed is a list of documents, when a single document is expected.

I agree, which is why we're prohibiting `PackedArray`, as we're always sure that it's an array, not an object. For PHP arrays however this isn't as simple and sometimes performance sensitive, which is why it wasn't added before (I'll leave it to @jmikola to confirm this). There is one more difference: with a structure like `[0 => 'foo', 1 => 'bar']`, the trained eye sees an array, but it can easily be used as an object. A `PackedArray` is by definition always an array, which is why we're prohibiting it.